File: mylang/go/internal/app/mylang/ident_func.go
package mylang

import "mylang/internal/app/mylang/lex"

type _IdentFunc struct {
	_Ident
	_Args      []*_Ident
	_ReturnLex _Lex
	_ReturnVar int
}

func MakeIdentFunc(name string, address int) _IdentFunc {
	newIdent := _IdentFunc{}
	newIdent._Name = name
	newIdent.putValue(address)
	newIdent.putType(lex.FUNCTION)
	return newIdent
}

func (ident *_IdentFunc) pushArg(arg *_Ident) {
	ident._Args = append(ident._Args, arg)
}

func (ident *_IdentFunc) sizeArgs() int {
	return len(ident._Args)
}

func (ident *_IdentFunc) setReturnLex(lx _Lex) {
	ident._ReturnLex = lx
}

func (ident *_IdentFunc) ReturnLex() _Lex {
	return ident._ReturnLex
}

func (ident *_IdentFunc) IArgs(i int) *_Ident {
	return ident._Args[i]
}

func (ident *_IdentFunc) setReturnVar(i int) {
	ident._ReturnVar = i
}

func (ident *_IdentFunc) ReturnVar() int {
	return ident._ReturnVar
}

File: mylang/rust/src/ident_func.rs
use crate::ident;
use crate::lex;
use std::ops::{Index, IndexMut};

pub struct IdentFunc {
    args: Vec<crate::ident::Ident>,
    return_lex: crate::lex::Lex,
    return_var: i32,
    ident: crate::ident::Ident,
}

impl IdentFunc {
    pub fn from_ident(ident: ident::Ident) -> IdentFunc {
        IdentFunc {
            args: Vec::new(),
            return_lex: lex::Lex::new(lex::Kind::NULL, 0, String::new()),
            return_var: 0,
            ident: ident,
        }
    }
    pub fn push_arg(&mut self, ident: ident::Ident) {
        self.args.push(ident.clone());
    }
    pub fn set_return_lex(&mut self, lex: lex::Lex) {
        self.return_lex = lex;
    }
    pub fn set_return_var(&mut self, i: i32) {
        self.return_var = i;
    }
    pub fn get_return_lex(&self) -> lex::Lex {
        self.return_lex.clone()
    }
    pub fn name(&self) -> &str {
        self.ident.name()
    }
    pub fn value(&self) -> i32 {
        self.ident.value()
    }
    pub fn kind(&self) -> crate::lex::Kind {
        self.ident.kind()
    }
    pub fn clone_args(&self) -> Vec<crate::ident::Ident> {
        self.args.clone()
    }
}

impl Index<usize> for IdentFunc {
    type Output = crate::ident::Ident;

    fn index(&self, index: usize) -> &Self::Output {
        &self.args[index]
    }
}

impl IndexMut<usize> for IdentFunc {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        &mut self.args[index]
    }
}
